### PR TITLE
EDS-19: Add SSE and failover

### PR DESCRIPTION
**Pull request by dlazarenko-macrometa**
Added `connectionTypes` - list of type connections to use in the connection. Order of values means priority of connection. For example, connectionTypes = ["ws", "sse"] means it will initially connect via WebSocket (ws).
 If unsuccessful, it will try Server-Sent Events (sse), and then loop back to retry WebSocket if needed.

```
let connection = PhotoniqEdsSdk.connect({
        host: "eds32-us-west.eng.macrometa.io",
        customerId: "cust-edsgdn",
        apiKey: "ptq_=***",
        fabric: "_system",
        connectionTypes: ["sse"],
    });
```

### CODE REVIEW

File: dist/connection-manager.d.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+ * Copyright (C) Macrometa, Inc - All Rights Reserved
+ *
+ * Unauthorized copying of this file, via any medium is strictly prohibited
+ * Proprietary and confidential
+ * Written by Macrometa, Inc <product@macrometa.com>, May 2024
+ */
+import { Config, EDSEvent, Connection, ConnectionStatus, ConnectionProperties, Filter } from "./types";
+import { QuerySet } from "./query-set";
+/**
+ * The main class manages connection and queries.
```
Comment (felipe-lopes-macrometa) [inline dist/connection-manager.d.ts:11]:
```suggestion
 * The main class manages the connection and queries.
```

File: dist/connection-manager.d.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    /**
+     * Get configuration of the connection
+     */
+    getConfig(): Config;
+    /**
+     * Get connection id
+     */
+    getId(): string | undefined;
+    /**
+     * Check weather it connected
```
Comment (felipe-lopes-macrometa) [inline dist/connection-manager.d.ts:40]:
```suggestion
     * Check the connection status
```

File: dist/connection-manager.js [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+ * Written by Macrometa, Inc <product@macrometa.com>, May 2024
+ */
+Object.defineProperty(exports, "__esModule", { value: true });
+exports.ConnectionManager = void 0;
+const types_1 = require("./types");
+const query_set_1 = require("./query-set");
+const switchable_connection_1 = require("./switchable-connection");
+const filters_state_1 = require("./filters-state");
+/**
+ * The main class manages connection and queries.
```
Comment (felipe-lopes-macrometa) [inline dist/connection-manager.js:16]:
```suggestion
 * The main class manages the connection and queries.
```

Comment (felipe-lopes-macrometa) [inline dist/connection-manager.js:16]: Maybe it is worth only having the documentation in the interface to not duplicate comments for the same methods

```suggestion
 * The main class manages the connection and queries.
```

File: dist/connection-manager.js [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+                                    self.handleErrorListenerForMap(querySetsAndCallbacksOnce, query, edsEvent);
+                                    self.handleGlobalListener(edsEvent);
+                                }
+                            }
+                        }
+                        if (querySetsAndCallbacksOnce.once) {
+                            self.queriesToQuerySetsAndCallbacks.delete(query);
+                            //console.log(`Once query deleted from queriesToQuerySetsAndCallbacks: ${query}`);
+                        }
+                    }*/
```
Comment (felipe-lopes-macrometa) [inline dist/connection-manager.js:109]: Remove unnecessary comments from published documentations
```suggestion
```

File: dist/connection-manager.js [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+        return this.config;
+    }
+    /**
+     * Get connection id
+     */
+    getId() {
+        return this.connection.getId();
+    }
+    /**
+     * Check weather it connected
```
Comment (felipe-lopes-macrometa) [inline dist/connection-manager.js:189]:
```suggestion
     * Check the connection status
```

File: dist/connection-manager.js [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+    status() {
+        if (this.connection) {
+            return this.connection.status();
+        }
+        else {
+            return types_1.ConnectionStatus.Closed;
+        }
+    }
+    handleGlobalListener(edsEvent) {
+        var _a;
```
Comment (felipe-lopes-macrometa) [inline dist/connection-manager.js:200]: Use let whenever possible in JS
```suggestion
        let _a;
```

File: dist/filters-state.js [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+            try {
+                callback(edsEvent);
+            }
+            catch (e) {
+                console.warn(`Error while handling error listener for query: ${query}`, e);
+            }
+        }
+    }
+    handleGlobalListener(edsEvent) {
+        var _a;
```
Comment (felipe-lopes-macrometa) [inline dist/filters-state.js:186]:
```suggestion
        let _a;
```

File: dist/sse/event-source.js [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+    }
+    onError(listener) {
+        this.errorListener = listener;
+    }
+    onClose(listener) {
+        this.closeListener = listener;
+    }
+    connect(data) {
+        return __awaiter(this, void 0, void 0, function* () {
+            var _a, _b, _c, _d, _e, _f;
```
Comment (felipe-lopes-macrometa) [inline dist/sse/event-source.js:33]:
```suggestion
            let _a, _b, _c, _d, _e, _f;
```

File: dist/switchable-connection.d.ts [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+ * Written by Macrometa, Inc <product@macrometa.com>, May 2024
+ */
+import { Config, ConnectionStatus, Filter, ConnectionProperties, EDSEvent, Connection } from "./types";
+import { QuerySet } from "./query-set";
+export declare class SwitchableConnection implements Connection {
+    private readonly config;
+    private readonly filtersState;
+    private connection?;
+    /**
+     * list of type connections to use in the conection. Order of values means priority of connection.
```
Comment (felipe-lopes-macrometa) [inline dist/switchable-connection.d.ts:15]:
```suggestion
     * List of type connections to use in the conection. Order of values means priority of connection.
```

File: dist/switchable-connection.d.ts [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+     * For example, connectionTypes = ["ws", "sse"] means it will initially connect via WebSocket (ws).
+     * If unsuccessful, it will try Server-Sent Events (sse), and then loop back to retry WebSocket if needed.
+     */
+    private connectionTypes;
+    /**
+     * value `-1` means it has not been connected yet or disconnected manually.
```
Comment (felipe-lopes-macrometa) [inline dist/switchable-connection.d.ts:21]:
```suggestion
     * Value `-1` means it has not been connected yet or disconnected manually.
```

File: dist/switchable-connection.d.ts [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+     * other values >= 0 mean the count of reconnections made before the established connection.
```
Comment (felipe-lopes-macrometa) [inline dist/switchable-connection.d.ts:22]:
```suggestion
     * Other values >= 0 mean the count of reconnections made before the established connection.
```

File: dist/switchable-connection.js [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+import { ConnectionStatus, EDSEventType } from "./types";
+import { WsConnection } from "./ws/ws-connection";
+import { SseConnection } from "./sse/sse-connection";
+import { FiltersState } from "./filters-state";
+import { QuerySet } from "./query-set";
+import { convertInitialData } from "./utils";
+export class SwitchableConnection {
+    constructor(config, globalListener) {
+        /**
+         * list of type connections to use in the conection. Order of values means priority of connection.
```
Comment (felipe-lopes-macrometa) [inline dist/switchable-connection.js:17]:
```suggestion
         * List of type connections to use in the conection. Order of values means priority of connection.
```

File: dist/switchable-connection.js [hunk 13]
```diff
@@ -0,0 +1,237 @@
...
+         * For example, connectionTypes = ["ws", "sse"] means it will initially connect via WebSocket (ws).
+         * If unsuccessful, it will try Server-Sent Events (sse), and then loop back to retry WebSocket if needed.
+         */
+        this.connectionTypes = ["ws"];
+        /**
+         * value `-1` means it has not been connected yet or disconnected manually.
```
Comment (felipe-lopes-macrometa) [inline dist/switchable-connection.js:23]:
```suggestion
         * Value `-1` means it has not been connected yet or disconnected manually.
```

File: dist/switchable-connection.js [hunk 14]
```diff
@@ -0,0 +1,237 @@
...
+         * other values >= 0 mean the count of reconnections made before the established connection.
```
Comment (felipe-lopes-macrometa) [inline dist/switchable-connection.js:24]:
```suggestion
         * Other values >= 0 mean the count of reconnections made before the established connection.
```

File: dist/switchable-connection.js [hunk 15]
```diff
@@ -0,0 +1,237 @@
...
+         */
+        this.reconnection = -1;
+        this.config = config;
+        this.filtersState = new FiltersState(globalListener);
+    }
+    /**
+     * Connect to Web Socket server
+     */
+    connect() {
+        var _a;
```
Comment (felipe-lopes-macrometa) [inline dist/switchable-connection.js:34]:
```suggestion
        let _a;
```

File: dist/switchable-connection.js [hunk 16]
```diff
@@ -0,0 +1,237 @@
...
+        (_a = this.connection) === null || _a === void 0 ? void 0 : _a.disconnect();
+    }
+    /**
+     * Get configuration of the connection
+     */
+    getConfig() {
+        return this.config;
+    }
+    /**
+     * Check weather it connected
```
Comment (felipe-lopes-macrometa) [inline dist/switchable-connection.js:198]:
```suggestion
     * Check the connection status
```

File: dist/ws/ws-connection.d.ts [hunk 17]
```diff
@@ -0,0 +1,33 @@
+import { Config, ConnectionProperties, ConnectionStatus, Filter, FilterState, InternalConnection } from "../types";
+import { FiltersState } from "../filters-state";
+export declare class WsConnection implements InternalConnection {
+    private readonly STUB_FILTER;
+    /**
+     * Default timeoput of ping-pong requests in seconds
```
Comment (felipe-lopes-macrometa) [inline dist/ws/ws-connection.d.ts:6]:
```suggestion
     * Default timeout of ping-pong requests in seconds
```

File: dist/ws/ws-connection.js [hunk 18]
```diff
@@ -0,0 +1,144 @@
+import { ConnectionStatus, PHOTONIQ_ES, } from "../types";
+import { tryToDecodeData } from "../utils";
+export class WsConnection {
+    constructor(config, filtersState) {
+        this.STUB_FILTER = "%7B%22action%22%3A%22remove%22%2C%22queries%22%3A%5B%22SELECT%20%2A%20FROM%20fake%22%5D%7D";
+        /**
+         * Default timeoput of ping-pong requests in seconds
```
Comment (felipe-lopes-macrometa) [inline dist/ws/ws-connection.js:7]:
```suggestion
         * Default timeout of ping-pong requests in seconds
```

File: src/switchable-connection.ts [hunk 19]
```diff
@@ -0,0 +1,262 @@
...
+import { QuerySet } from "./query-set";
+import { convertInitialData } from "./utils";
+
+export class SwitchableConnection implements Connection {
+
+    private readonly config: Config;
+    private readonly filtersState: FiltersState;
+    private connection?: InternalConnection;
+    /**
+     * list of type connections to use in the conection. Order of values means priority of connection.
```
Comment (felipe-lopes-macrometa) [inline src/switchable-connection.ts:34]:
```suggestion
     * List of type connections to use in the conection. Order of values means priority of connection.
```

File: src/switchable-connection.ts [hunk 20]
```diff
@@ -0,0 +1,262 @@
...
+     * For example, connectionTypes = ["ws", "sse"] means it will initially connect via WebSocket (ws).
+     * If unsuccessful, it will try Server-Sent Events (sse), and then loop back to retry WebSocket if needed.
+     */
+    private connectionTypes = ["ws"];
+    /**
+     * value `-1` means it has not been connected yet or disconnected manually.
+     * other values >= 0 mean the count of reconnections made before the established connection.
```
Comment (felipe-lopes-macrometa) [inline src/switchable-connection.ts:41]:
```suggestion
     * Value `-1` means it has not been connected yet or disconnected manually.
     * Other values >= 0 mean the count of reconnections made before the established connection.
```

File: src/switchable-connection.ts [hunk 21]
```diff
@@ -0,0 +1,262 @@
...
+    
+    /**
+     * Get configuration of the connection
+     */
+    public getConfig(): Config {
+        return this.config;
+    }
+
+    /**
+     * Check weather it connected
```
Comment (felipe-lopes-macrometa) [inline src/switchable-connection.ts:223]:
```suggestion
     * Check the connection status
```

File: src/utils.ts [hunk 22]
```diff
@@ -0,0 +1,64 @@
...
+    for (let sqlParameter in sqlData) {
+        let path = sqlParameter.split('.');
+        if (path.length <= 1) {
+            continue;
+        }
+        let value = sqlData;
+        for (let i = 0; i < path.length; i++) {
+        if (value[path[i]] === undefined) {
+            value[path[i]] = {};
+        }
```
Comment (felipe-lopes-macrometa) [inline src/utils.ts:29]:
```suggestion
            if (value[path[i]] === undefined) {
                value[path[i]] = {};
            }
```

File: src/ws/ws-connection.ts [hunk 23]
```diff
@@ -0,0 +1,180 @@
...
+    PHOTONIQ_ES,
+} from "../types";
+import { FiltersState } from "../filters-state";
+import { tryToDecodeData } from "../utils";
+
+export class WsConnection implements InternalConnection {
+    
+    private readonly STUB_FILTER: string = "%7B%22action%22%3A%22remove%22%2C%22queries%22%3A%5B%22SELECT%20%2A%20FROM%20fake%22%5D%7D";
+    /**
+     * Default timeoput of ping-pong requests in seconds
```
Comment (felipe-lopes-macrometa) [inline src/ws/ws-connection.ts:17]:
```suggestion
     * Default timeout of ping-pong requests in seconds 
```

File: dist/connection-manager.js (hunk 3 above)
Comment (dlazarenko-macrometa) [reply dist/connection-manager.js:16]: This file is not using at all, deleted

File: dist/connection-manager.d.ts (hunk 2 above)
Comment (dlazarenko-macrometa) [reply dist/connection-manager.d.ts:40]: anything in `dist` is auto-generated, fixed in `src`

File: dist/connection-manager.js (hunk 4 above)
Comment (dlazarenko-macrometa) [reply dist/connection-manager.js:109]: This file is not using at all, deleted

File: dist/connection-manager.d.ts (hunk 1 above)
Comment (dlazarenko-macrometa) [reply dist/connection-manager.d.ts:11]: This file is not using at all, deleted

File: dist/connection-manager.js (hunk 5 above)
Comment (dlazarenko-macrometa) [reply dist/connection-manager.js:189]: This file is not using at all, deleted

File: dist/connection-manager.js (hunk 6 above)
Comment (dlazarenko-macrometa) [reply dist/connection-manager.js:200]: This file is not using at all, deleted

File: dist/sse/event-source.js (hunk 8 above)
Comment (dlazarenko-macrometa) [reply dist/sse/event-source.js:33]: I cannot change it as it is autogenerated

File: dist/switchable-connection.d.ts (hunk 9 above)
Comment (dlazarenko-macrometa) [reply dist/switchable-connection.d.ts:15]: Done in `switchable-connection.ts`!

File: dist/switchable-connection.d.ts (hunk 10 above)
Comment (dlazarenko-macrometa) [reply dist/switchable-connection.d.ts:21]: Done in `switchable-connection.ts`!

File: dist/switchable-connection.d.ts (hunk 11 above)
Comment (dlazarenko-macrometa) [reply dist/switchable-connection.d.ts:22]: Done in `switchable-connection.ts`!

File: dist/switchable-connection.js (hunk 12 above)
Comment (dlazarenko-macrometa) [reply dist/switchable-connection.js:17]: Done in `switchable-connection.ts`!

File: dist/switchable-connection.js (hunk 13 above)
Comment (dlazarenko-macrometa) [reply dist/switchable-connection.js:23]: Done in `switchable-connection.ts`!

File: dist/switchable-connection.js (hunk 14 above)
Comment (dlazarenko-macrometa) [reply dist/switchable-connection.js:24]: Done in `switchable-connection.ts`!

File: dist/switchable-connection.js (hunk 16 above)
Comment (dlazarenko-macrometa) [reply dist/switchable-connection.js:204]: Done in `switchable-connection.ts`!

File: dist/ws/ws-connection.js (hunk 18 above)
Comment (dlazarenko-macrometa) [reply dist/ws/ws-connection.js:7]: Done in `ws-connection.ts`!

File: dist/ws/ws-connection.d.ts (hunk 17 above)
Comment (dlazarenko-macrometa) [reply dist/ws/ws-connection.d.ts:6]: Done in `ws-connection.ts`!

File: src/switchable-connection.ts (hunk 19 above)
Comment (dlazarenko-macrometa) [reply src/switchable-connection.ts:34]: Done!

File: src/switchable-connection.ts (hunk 20 above)
Comment (dlazarenko-macrometa) [reply src/switchable-connection.ts:41]: Done!

File: src/utils.ts (hunk 22 above)
Comment (dlazarenko-macrometa) [reply src/utils.ts:29]: Done!

File: src/ws/ws-connection.ts (hunk 23 above)
Comment (dlazarenko-macrometa) [reply src/ws/ws-connection.ts:17]: Done!

**Review by felipe-lopes-macrometa — APPROVED** [review]

**Pull request merged**